File: Editorial/0098_Validate_Binary_Search_Tree/Go/1_Recursive_Traversal_with_Valid_Range.go
func validate(root *TreeNode, low *int, high *int) bool {
    // Empty trees are valid BSTs.
    if root == nil {
        return true
    }
    // The current node's value must be between low and high.
    if (low != nil && root.Val <= *low) || (high != nil && root.Val >= *high) {
        return false
    }
    // The left and right subtree must also be valid.
    return validate(root.Right, &root.Val, high) &&
        validate(root.Left, low, &root.Val)
}

func isValidBST(root *TreeNode) bool {
    return validate(root, nil, nil)
}

File: Editorial/0098_Validate_Binary_Search_Tree/TypeScript/1_Recursive_Traversal_with_Valid_Range.ts
function isValidBST(
    root: TreeNode | null,
    low: number | null = null,
    high: number | null = null,
): boolean {
    // Empty trees are valid BSTs.
    if (root === null) return true;
    // The current node's value must be between low and high.
    if (
        (low !== null && root.val <= low) ||
        (high !== null && root.val >= high)
    )
        return false;
    // The left and right subtree must also be valid.
    return (
        isValidBST(root.right, root.val, high) &&
        isValidBST(root.left, low, root.val)
    );
}
